src/libraries/System.Text.Json/tests/Serialization/ExtensionDataTests.cs

```diff
@@ -220,6 +221,12 @@ private class ClassWithTwoExtensionProperties
             public Dictionary<string, object> MyOverflow2 { get; set; }
         }
 
+        private class ClassWithNotSupportedExtensionPropertyImmutableDictionary
```

Comment:
We should test other classes like `IImmutableDictionary<string, JsonElement>`, and a custom dictionary without a public parameterless ctor.

Reply:
Can do.

Reply:
I've added additional collection classes and tests for these cases.

When trying to test for these cases I noticed that classes that implement `IDictionary<TKey, TValue>` but do not have two generic type parameters end up throwing `IndexOutOfRangeException` on de-serialization when marked with `[JsonExtensionData]`.  Should this also be a `NotSupportedException` instead?

For example, by using the `StringToGenericIDictionaryWrapper<TValue>` class already in this test project, the following simple test will reproduce the issue:

```
[Fact]
public static void DeserializeIntoSingleGenericDictionary()
{
    JsonSerializer.Deserialize<ClassWithExtensionPropertyStringToGeneric>("{\"hello\":\"world\"}");
}

private class ClassWithExtensionPropertyStringToGeneric
{
    [JsonExtensionData]
    public StringToGenericIDictionaryWrapper<object> MyOverflow { get; set; }
}
```

Reply:
Ah, yes. Thes